File: src/minimum_absolute_difference_between_elements_with_constraint_2817/solution_two_heaps.go
package minimum_absolute_difference_between_elements_with_constraint_2817

import (
	"container/heap"
	"fmt"
	"math"
	"sort"
)

type Pair struct {
	idx int
	val int
}

type MinHeap []Pair

func (h MinHeap) Len() int {
	return len(h)
}

func (h MinHeap) Less(i, j int) bool {
	return h[i].idx < h[j].idx
}

func (h MinHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h *MinHeap) Push(x any) {
	*h = append(*h, x.(Pair))
}

func (h *MinHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func minAbsoluteDifferenceHeap(nums []int, x int) int {
	sortedList := make([]Pair, len(nums))
	for i, val := range nums {
		sortedList[i] = Pair{idx: i, val: val}
	}
	sort.Slice(sortedList, func(i, j int) bool {
		if sortedList[i].val == sortedList[j].val {
			return sortedList[i].idx < sortedList[j].idx
		}
		return sortedList[i].val < sortedList[j].val
	})

	minHeap := &MinHeap{}
	maxHeap := &MinHeap{}

	result := math.MaxInt
	for _, pair := range sortedList {
		val, idx := pair.val, pair.idx
		heap.Push(minHeap, Pair{idx: idx, val: val})
		heap.Push(maxHeap, Pair{idx: -idx, val: val})

		for minHeap.Len() > 0 && (*minHeap)[0].idx <= idx-x {
			top := heap.Pop(minHeap).(Pair)
			result = min(result, val-top.val)
		}

		for maxHeap.Len() > 0 && (*maxHeap)[0].idx <= -idx-x {
			top := heap.Pop(maxHeap).(Pair)
			result = min(result, val-top.val)
		}
	}

	return result
}

func TestHeap() {
	fmt.Println(minAbsoluteDifferenceHeap([]int{4, 3, 2, 4}, 2))
	fmt.Println(minAbsoluteDifferenceHeap([]int{5, 3, 2, 10, 15}, 1))
	fmt.Println(minAbsoluteDifferenceHeap([]int{1, 2, 3, 4}, 3))
	fmt.Println(minAbsoluteDifferenceHeap([]int{14, 111, 16}, 1))
}

File: src/minimum_absolute_difference_between_elements_with_constraint_2817/solution_two_heaps.ts
class MinHeap<T> {
  private heap: T[] = [];

  push(item: T): void {
    this.heap.push(item);
    this.bubbleUp();
  }

  pop(): T | undefined {
    if (this.size() === 0) return undefined;
    this.swap(0, this.heap.length - 1);
    const item = this.heap.pop();
    this.bubbleDown();
    return item;
  }

  peek(): T | undefined {
    return this.heap[0];
  }

  size(): number {
    return this.heap.length;
  }

  private bubbleUp(): void {
    let index = this.heap.length - 1;
    const current = this.heap[index];

    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      const parent = this.heap[parentIndex];
      if (this.compare(current, parent) >= 0) break;

      this.swap(index, parentIndex);
      index = parentIndex;
    }
  }

  private bubbleDown(): void {
    let index = 0;

    while (true) {
      const leftIndex = 2 * index + 1;
      const rightIndex = 2 * index + 2;
      let smallestIndex = index;

      if (
        leftIndex < this.heap.length &&
        this.compare(this.heap[leftIndex], this.heap[smallestIndex]) < 0
      ) {
        smallestIndex = leftIndex;
      }
      if (
        rightIndex < this.heap.length &&
        this.compare(this.heap[rightIndex], this.heap[smallestIndex]) < 0
      ) {
        smallestIndex = rightIndex;
      }
      if (smallestIndex === index) break;

      this.swap(index, smallestIndex);
      index = smallestIndex;
    }
  }

  private swap(i: number, j: number): void {
    [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
  }

  protected compare(a: T, b: T): number {
    return (a as any) - (b as any); // Override this for custom comparisons
  }
}

class MinHeapWithCustomComparator<T> extends MinHeap<[number, T]> {
  protected compare(a: [number, T], b: [number, T]): number {
    return a[0] - b[0];
  }
}

class MaxHeapWithCustomComparator<T> extends MinHeap<[number, T]> {
  protected compare(a: [number, T], b: [number, T]): number {
    return b[0] - a[0];
  }
}

function minDiff(nums: number[], x: number): number {
  const sortedList: [number, number][] = nums
    .map<[number, number]>((val, idx) => [val, idx])
    .sort((a, b) => a[0] - b[0]);

  const minHeap = new MinHeapWithCustomComparator<number>();
  const maxHeap = new MaxHeapWithCustomComparator<number>();
  let result = Infinity;

  for (const [val, idx] of sortedList) {
    minHeap.push([idx, val]);
    maxHeap.push([idx, val]);

    while (minHeap.size() > 0 && minHeap.peek()![0] <= idx - x) {
      result = Math.min(result, Math.abs(val - minHeap.pop()![1]));
    }

    while (maxHeap.size() > 0 && maxHeap.peek()![0] >= idx + x) {
      result = Math.min(result, Math.abs(val - maxHeap.pop()![1]));
    }
  }

  return result;
}

export function test() {
  console.log(minDiff([4, 3, 2, 4], 2));
  console.log(minDiff([5, 3, 2, 10, 15], 1));
  console.log(minDiff([1, 2, 3, 4], 3));
  console.log(minDiff([14, 111, 16], 1));
}
